### PR TITLE
[Feature] made a classifier based on information retrieved from QR on the omr sheets

### PR DESCRIPTION
**Pull request by rudrapsc**
Fixes https://github.com/Udayraj123/OMRChecker/issues/105

### CODE REVIEW

**Comment by Udayraj123** [comment]
Change the base branch to dev for now

**Review by Udayraj123 — COMMENTED** [review]
In process

File: src/processors/CropPage.py [hunk 1]
```diff
@@ -55,9 +55,8 @@ def angle(p_1, p_2, p_0):
 class CropPage(ImagePreprocessor):
     def __init__(self, cropping_ops, args):
         self.args = args
-        self.morph_kernel = tuple(
-            int(x) for x in cropping_ops.get("morphKernel", [10, 10])
```
Comment (Udayraj123) [inline src/processors/CropPage.py:59]: use `map(int, <your list>)`

File: src/defaults/schema.json [hunk 2]
```diff
@@ -274,7 +295,8 @@
                   "type": "object",
                   "propertyNames": {
                     "enum": [
-                      "morphKernel"
+                      "morphKernel",
+                      "Use_Read_Barcode"
```
Comment (Udayraj123) [inline src/defaults/schema.json:299]: use_read_barcode

File: src/processors/ReadBarcode.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        self.top_right = options.get("top_right")
+        self.qr_to_output = options.get("qr_to_output_directory")
+        self.input_sorting = options.get("input_sorting", False)
+        self.y1 = self.top_left[1]
+        self.y2 = self.top_right[1]
+        self.x1 = self.top_left[0]
+        self.x2 = self.top_right[1]
+
+
+    def detect(image):
```
Comment (Udayraj123) [inline src/processors/ReadBarcode.py:21]: add `@staticmethod` decorator for this function as no `self` is passed

File: src/processors/ReadBarcode.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+            logger.error(
+                "\tError: QR not found :Have you accidentally included ReadBarcode plugin?"
+            )
+        return data
+
+    def apply_filter(self, img, args):
+        img1 = img[self.y1 : self.y2, self.x1 : self.x2]
+
+        data = ReadBarcode.detect(img1)
+        if self.qr_to_output !={}:
```
Comment (Udayraj123) [inline src/processors/ReadBarcode.py:40]: run pre-commit once again

File: src/utils/TemplateByBarcode.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+from src.logger import logger
+from src.processors.manager import ProcessorManager
+from src.template import Template
+from src.utils.imgutils import ImageUtils
+
+# import cv2
+
+
+class TemplateByBarcode:
+    def update_template(local_template_path, path, args, PROCESSOR_MANAGER,curr_dir,root_dir):
```
Comment (Udayraj123) [inline src/utils/TemplateByBarcode.py:15]: add `@staticmethod` decorators

File: src/utils/TemplateByBarcode.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+                    config.dimensions.processing_height,
+                )
+
+        for pre_processor in template.TemplateByBarcode:
+            data, input_sorting = pre_processor.apply_filter(in_omr, args)
+            path = data
+            path_input = out.paths.output_dir
+            path_1 = str(save_dir[:-1])
+            TemplateByBarcode.make_folders(path_1, data[:-1])
+            data_name = f"configs/{str(data[:-1])}-Template.json"
```
Comment (Udayraj123) [inline src/utils/TemplateByBarcode.py:52]: `-template.json`

File: src/utils/TemplateByBarcode.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+            local_template_path = root_dir.joinpath(data_name)
+            if os.path.exists(local_template_path):
+                template, out = TemplateByBarcode.update_template(
+                    local_template_path, path, args, PROCESSOR_MANAGER,curr_dir,root_dir
+                )
+            else:
+                logger.error(f"Unable to find the path {local_template_path}")
+                return None, None
+            if input_sorting:
+                data_2 = f"{data[:-1]}_inputs"
```
Comment (Udayraj123) [inline src/utils/TemplateByBarcode.py:62]: What is the suffix for?

Comment (rudrapsc) [reply src/utils/TemplateByBarcode.py]: this is to create folder where sorted input images will be saved

**Review by Udayraj123 — DISMISSED** [review]
LGTM

**Comment by rudrapsc** [comment]
to check the program
follow the steps in Readme:-
then run `python3 main.py -i ./samples/sample7`

**Comment by Udayraj123** [comment]
@rudrapsc can you share the proposal link here as well? It'll help new folks to go through the process for new feature work

**Comment by Udayraj123** [comment]
Also for master merge, you'd need a more generic approach as discussed. On top of the evaluation PR (which contains a revamp). https://github.com/Udayraj123/OMRChecker/pull/118

**Comment by Udayraj123** [comment]
For master merge we will need to make the code compatible with https://github.com/Udayraj123/OMRChecker/pull/118 getting merged now

**Comment by rudrapsc** [comment]
> For master merge we will need to make the code compatible with #118 getting merged now

I am busy with some other work as of now, can I get back on this by 24th feb'23?

**Comment by Udayraj123** [comment]
Cool that is alright @rudrapsc. You can ping back by the month end.

**Comment by Udayraj123** [comment]
Thought of a more generalised version and will write a schema spec for it on the github issue

**Pull request merged**